node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+						// query basic system parameters once
+						// @todo assumption: these cannot change within a session
+						let (validator_set, signing_context) =
+							query_basics(&mut ctx, relay_parent).await?;
+
+						let _ = tracker.per_relay_parent.insert(
+							relay_parent,
+							PerRelayParentData {
+								signing_context,
+								validator_set,
+								..Default::default()
+							},
+						);
+					}
+					FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+						trace!(target: "bitd", "Stop {:?}", relay_parent);
+						// @todo assumption: it is good enough to prevent additional work from being
+						// scheduled, the individual futures are supposedly completed quickly
+						let _ = tracker.per_relay_parent.remove(&relay_parent);
+					}
+					FromOverseer::Signal(OverseerSignal::Conclude) => {
+						trace!(target: "bitd", "Conclude");
+						tracker.per_relay_parent.clear();
+						return Ok(());
+					}
+				}
+			}
+		}
+	}
+}
+
+/// Modify the reputation of peer based on their behaviour.
+async fn modify_reputation<Context>(
+	ctx: &mut Context,
+	peer: PeerId,
+	rep: ReputationChange,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	trace!(target: "bitd", "Reputation change of {:?} for peer {:?}", rep, peer);
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::ReportPeer(peer, rep),
+	))
+	.await
+}
+
+/// Distribute a given valid and signature checked bitfield message.
+///
+/// Can be originated by another subsystem or received via network from another peer.
+async fn relay_message<Context>(
+	ctx: &mut Context,
+	job_data: &mut PerRelayParentData,
+	peer_views: &mut HashMap<PeerId, View>,
+	validator: ValidatorId,
+	message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+
+	// notify the overseer about a new and valid signed bitfield
+	ctx.send_message(
+		AllMessages::Provisioner(
+			ProvisionerMessage::ProvisionableData(
+				ProvisionableData::Bitfield(
+					message.relay_parent.clone(),
+					message.signed_availability.clone(),
+				)
+			)
+		)
+	).await;
+
+	let message_sent_to_peer = &mut (job_data.message_sent_to_peer);
+
+	// concurrently pass on the bitfield distribution to all interested peers
+	let interested_peers = peer_views
+		.iter()
+		.filter_map(|(peer, view)| {
+			// check interest in the peer in this message's relay parent
+			if view.contains(&message.relay_parent) {
+				// track the message as sent for this peer
+				message_sent_to_peer
+					.entry(peer.clone())
+					.or_default()
+					.insert(validator.clone());
+
+				Some(peer.clone())
+			} else {
+				None
+			}
+		})
+		.collect::<Vec<PeerId>>();
+
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::SendMessage(
+			interested_peers,
+			BitfieldDistribution::PROTOCOL_ID,
+			Encode::encode(&message),
```

Comment:
Typically we use

```suggestion
			message.encode(),
```

Any particular reason for favoring the UFCS version here?

Reply:
False positive change of a rustc error about missing traits, UFCS  avoids ambiguities (sometimes) and I have no preference.